PolyEngine/Core/Src/Collections/BitMask.cpp

```diff
@@ -0,0 +1,382 @@
+
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		BitMask temp(BitsNumber);
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] | rhs.BitList[i];
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i];
+		return temp;
+	}
+
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] | rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = rhs.BitList[i];
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask BitMask::operator^(const BitMask& rhs) const
+{
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		size_t tempBitsNumber = 0;
+		if (BitsNumber > rhs.BitsNumber)
+			tempBitsNumber = BitsNumber;
+		else
+			tempBitsNumber = rhs.BitsNumber;
+
+		BitMask temp(tempBitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] ^ rhs.BitList[i];
+		return temp;
+	}
+	//Lhs size bigger than rhs size
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		BitMask temp(BitsNumber);
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] ^ rhs.BitList[i];
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i];
+		return temp;
+	}
+	//rhs size bigger than lhs size
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] ^ rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = rhs.BitList[i];
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask BitMask::operator&(const BitMask& rhs) const
+{
+	if (BitList.GetSize() == rhs.BitList.GetSize())
+	{
+		size_t tempBitsNumber = 0;
+		if (BitsNumber > rhs.BitsNumber)
+			tempBitsNumber = BitsNumber;
+		else
+			tempBitsNumber = rhs.BitsNumber;
+
+		BitMask temp(tempBitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		return temp;
+	}
+	else if (BitList.GetSize() > rhs.BitList.GetSize())
+	{
+		BitMask temp(BitsNumber);
+		for (size_t i = 0; i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		for (size_t i = rhs.BitList.GetSize(); i < BitList.GetSize(); i++)
+			temp.BitList[i] = ZERO;
+		return temp;
+	}
+	else if (BitList.GetSize() < rhs.BitList.GetSize())
+	{
+		BitMask temp(rhs.BitsNumber);
+		for (size_t i = 0; i < BitList.GetSize(); i++)
+			temp.BitList[i] = BitList[i] & rhs.BitList[i];
+		for (size_t i = BitList.GetSize(); i < rhs.BitList.GetSize(); i++)
+			temp.BitList[i] = ZERO;
+		return temp;
+	}
+	return BitMask(0);
+}
+
+BitMask& BitMask::operator~()
+{
+	for (auto& x : BitList)
+		x = ~x;
+
+	return *this;
+}
+
+
+bool BitMask::Resize(const int offset)
```

Comment:
Resize should apply given size, not current size + offset. That way this method is simple 
`BitsNumber = size; BitList.Resize((size + TYPE_BIT - 1)/TYPE_BIT)`